Extensions/AuthPropertiesTokenCacheHelper.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.Security.Claims;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.Cookies;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Identity.Client;
+
+namespace TodoListService.Extensions
+{
+    public class AuthPropertiesTokenCacheHelper
+    {
+        private const string TokenCacheKey = ".UserTokenCache";
+
+        private HttpContext _httpContext;
+        private ClaimsPrincipal _principal;
+        private AuthenticationProperties _authProperties;
+        private string _signInScheme;
+        public TokenCache TokenCache { get; private set; }
+
+        private AuthPropertiesTokenCacheHelper(AuthenticationProperties authProperties) : base()
+        {
+            _authProperties = authProperties;
+            TokenCache = new TokenCache();
+            TokenCache.SetBeforeAccess(BeforeAccessNotificationWithProperties);
+            TokenCache.SetAfterAccess(AfterAccessNotificationWithProperties);
+            TokenCache.SetBeforeWrite(BeforeWriteNotification);
+        }
+
+        private AuthPropertiesTokenCacheHelper(HttpContext httpContext, string signInScheme) : base()
+        {
+            _httpContext = httpContext;
+            _signInScheme = signInScheme;
+            TokenCache = new TokenCache();
+            TokenCache.SetBeforeAccess(BeforeAccessNotificationWithContext);
+            TokenCache.SetAfterAccess(AfterAccessNotificationWithContext);
+            TokenCache.SetBeforeWrite(BeforeWriteNotification);
+        }
+
+        public static TokenCache ForCodeRedemption(AuthenticationProperties authProperties)
+        {
+            return new AuthPropertiesTokenCacheHelper(authProperties).TokenCache;
+        }
+
+        public static TokenCache ForApiCalls(HttpContext httpContext,
+            string signInScheme = CookieAuthenticationDefaults.AuthenticationScheme)
+        {
+            return new AuthPropertiesTokenCacheHelper(httpContext, signInScheme).TokenCache;
+        }
+
+        private void BeforeAccessNotificationWithProperties(TokenCacheNotificationArgs args)
+        {
+            string cachedTokensText;
+            if (_authProperties.Items.TryGetValue(TokenCacheKey, out cachedTokensText))
+            {
+                var cachedTokens = Convert.FromBase64String(cachedTokensText);
+                TokenCache.Deserialize(cachedTokens);
```

Comment:
I don't think there's any need to have a 'cachedTokens' variable. This could just be 'TokenCache.Deserialize(Convert.FromBase64String(cachedTokensText))'.

Then since it's all on one line it'll also be possible to remove the braces from the 'if' statement.

Reply:
Made this change and few other similar ones.